added user given output directory

tdincer opened the pull request:
Added:
1. user given output directory option to `cnmf.fit_file` method
2. option for `cnmf.fit_file` to return movie correction object

tdincer committed:
added user given output directory

tdincer commented:
This modified pipeline was successfully run on a sample dataset. Let's try to push this to the `flatironinstitute/caiman` fork before Friday. Any feedback is appreciated, especially on backward compatibility.

jverswijver commented on caiman/motion_correction.py:
```diff
@@ -2829,14 +2837,14 @@ def motion_correct_batch_rigid(fname, max_shifts, dview=None, splits=56, num_spl
             base_name=os.path.split(fname[0])[-1][:-4] + '_rig_'
         else:
-            base_name=os.path.split(fname)[-1][:-4] + '_rig_'
+            base_name=os.path.split(fname)[-1][:-4] + '_rig_'   
```
```suggestion
            base_name=os.path.split(fname)[-1][:-4] + '_rig_'
```

jverswijver replied:
Don't leave whitespace at the end of lines.

tdincer replied:
Thank you @jverswijver!

jverswijver reviewed (changes requested):
Looks great @tdincer! just one minor change with some whitespace at the end of a line. :+1:

tdincer committed:
Update caiman/motion_correction.py

Co-authored-by: jverswijver <49455164+jverswijver@users.noreply.github.com>

ttngu207 commented on caiman/motion_correction.py:
```diff
@@ -173,15 +174,19 @@ def __init__(self, fname, min_mov=None, dview=None, max_shifts=(6, 6), niter_rig
-            logging.info('Creating file for motion correction "tmp_mov_mot_corr.hdf5"')
-            cm.movie(fname).save('tmp_mov_mot_corr.hdf5')
-            fname = ['tmp_mov_mot_corr.hdf5']
+            logging.info(f'Creating file for motion correction "{pathlib.Path(output_dir).joinpath("tmp_mov_mot_corr.hdf5")}"')
+            cm.movie(fname).save(f'{output_dir}/tmp_mov_mot_corr.hdf5')
+            fname = [pathlib.Path(output_dir).joinpath("tmp_mov_mot_corr.hdf5").as_posix()]
```
```suggestion
            tmp_mov_mot_corr_fp = pathlib.Path(output_dir) / 'tmp_mov_mot_corr.hdf5'
            logging.info(f'Creating file for motion correction "{tmp_mov_mot_corr_fp }"')
            cm.movie(fname).save(tmp_mov_mot_corr_fp.as_posix())
            fname = [tmp_mov_mot_corr_fp.as_posix()]
```

ttngu207 commented on caiman/source_extraction/cnmf/cnmf.py:
```diff
@@ -324,8 +324,13 @@ def fit_file(self, motion_correct=False, indices=None, include_eval=False):
                 flag for performing component evaluation
+            output_dir (str)
+                directory to save the outputs
+            return_mc (bool)
+                flag to return motion correction object
```
We should safeguard against users setting `return_mc=True` but `motion_correct=False`, conflicting input values.

Perhaps with something like this, somehwere
```
return_mc = return_mc if motion_correct else False
```

tdincer replied:
Good point @ttngu207! I'll fix this conditional issue.

tdincer replied:
@ttngu207 Line 400 fixes this issue

ttngu207 commented on caiman/source_extraction/cnmf/cnmf.py:
```diff
@@ -337,7 +342,7 @@ def fit_file(self, motion_correct=False, indices=None, include_eval=False):
             raise Exception('File not found!')
 
-        base_name = pathlib.Path(fnames[0]).stem + "_memmap_"
+        base_name = (pathlib.Path(output_dir) / (pathlib.Path(fnames[0]).stem + "_memmap_")).as_posix()
```
I'd double check here on how `base_name` is used. I think `base_name` should not contain any path, but only a filename (or prefix to a filename), so we should not join `output_dir` here.

tdincer replied:
@ttngu207  `base_name` is passed only to `mmapping.save_memmap` together with the full path file names in line 369 or 372, depending on a condition. Addition of another path related keyword looks excessive to me. There is only one caveat. The docstring in `mmapping.save_memmap` clearly states that the `base_name` must not contain underscore character. As long as this caveat is considered we should be good.

ttngu207 replied:
We cannot guarantee `base_name` will not contain underscore especially when `base_name` is now a full path.
On second look, I do think you don't need to include the `output_dir` as part of the `base_name` here. And you don't need to pass `output_dir` into `mmapping.save_memmap`.
In the `mmapping.save_memmap`, the path is actually retrieved from `fname`, which, in this case, is passed in as `fname_mc` (line 369 below).
`fname_mc` is generated from the output of `mc.motion_correct`, which should already account for the correct `output_dir`.

ttngu207 reviewed (changes requested):
overall looks good, I have a few comments/suggestions

tdincer committed:
Update caiman/motion_correction.py

Co-authored-by: Thinh Nguyen <thinhnguyen0405@gmail.com>

tdincer committed:
Update cnmf.py

changed `if return_mc:` to ` if return_mc & motion_correct:` to make sure that motion correction is applied if we want to return the `motion correction object`

tdincer committed:
Update cnmf.py

`base_name` line is changed to its original form.

tdincer committed:
adjusted .npz file output directory

tdincer commented:
@ttngu207 This one is finally ready to merge (see [adjusted .npz file output directory](https://github.com/vathes/CaImAn/pull/1/commits/4ee1ff99a8143e76edc6a5f2795e716b507201d0)).
